atc/metric/emitter/influxdb.go

```diff
@@ -46,48 +58,83 @@ func (config *InfluxDBConfig) NewEmitter() (metric.Emitter, error) {
 	}
 
 	return &InfluxDBEmitter{
-		client:   client,
-		database: config.Database,
+		Client:   	client,
+		Database: 	config.Database,
+		BatchSize: 	int(config.BatchSize),
+		BatchDuration: 	config.BatchDuration,
 	}, nil
 }
 
-func (emitter *InfluxDBEmitter) Emit(logger lager.Logger, event metric.Event) {
+func emitBatch(emitter *InfluxDBEmitter, logger lager.Logger, events []metric.Event) {
+
+	logger.Debug("influxdb-emit-batch", lager.Data{
+		"size": len(events),
+	})
 	bp, err := influxclient.NewBatchPoints(influxclient.BatchPointsConfig{
-		Database: emitter.database,
+		Database: emitter.Database,
 	})
 	if err != nil {
 		logger.Error("failed-to-construct-batch-points", err)
 		return
 	}
 
-	tags := map[string]string{
-		"host": event.Host,
-	}
-
-	for k, v := range event.Attributes {
-		tags[k] = v
+	for _, event := range events {
+		tags := map[string]string{
+			"host": event.Host,
+		}
+
+		for k, v := range event.Attributes {
+			tags[k] = v
+		}
+
+		point, err := influxclient.NewPoint(
+			event.Name,
+			tags,
+			map[string]interface{}{
+				"value": event.Value,
+				"state": string(event.State),
+			},
+			event.Time,
+		)
+		if err != nil {
+			logger.Error("failed-to-construct-point", err)
+			continue
+		}
+
+		bp.AddPoint(point)
 	}
 
-	point, err := influxclient.NewPoint(
-		event.Name,
-		tags,
-		map[string]interface{}{
-			"value": event.Value,
-			"state": string(event.State),
-		},
-		event.Time,
-	)
+	err = emitter.Client.Write(bp)
 	if err != nil {
-		logger.Error("failed-to-construct-point", err)
+		logger.Error("failed-to-send-points",
+			errors.Wrap(metric.ErrFailedToEmit, err.Error()))
 		return
 	}
+	logger.Info("influxdb-emitter-fork-influxdb-batch-emitted", lager.Data{
+		"size": len(events),
+	})
+}
 
-	bp.AddPoint(point)
 
-	err = emitter.client.Write(bp)
-	if err != nil {
-		logger.Error("failed-to-send-points",
-			errors.Wrap(metric.ErrFailedToEmit, err.Error()))
-		return
+func (emitter *InfluxDBEmitter) Emit(logger lager.Logger, event metric.Event) {
+	batch = append(batch, event)
+	duration := time.Since(lastBatchTime)
+	if len(batch) >= emitter.BatchSize || duration >= emitter.BatchDuration {
+		logger.Debug("influxdb-pre-emit-batch", lager.Data{
+			"influxdb-batch-size": emitter.BatchSize,
+			"current-batch-size": len(batch),
+			"influxdb-batch-duration": emitter.BatchDuration,
+			"current-duration": duration,
+		})
+		emitter.SubmitBatch(logger)
 	}
 }
+
+func (emitter *InfluxDBEmitter) SubmitBatch(logger lager.Logger) {
+	batchToSubmit := make([]metric.Event, len(batch))
+	copy(batchToSubmit, batch)
+	batch = make([]metric.Event, 0)
+	lastBatchTime = time.Now()
+	go emitBatch(emitter, logger, batchToSubmit)
```

Comment:
I guess the idea behind this is to prevent a giant batch from submitting too slowly and causing the queue to fill up. Makes sense, since that was the original problem. But I wonder if batching would help mitigate that in and of itself, since submiting 5k at a time might cause less backpressure than 1 event 5000 times.

Have you observed any issues with this yet at large scale? It seems like the goroutines could potentially pile up due to a slow consumer.

I think @cirocosta had similar concerns on an earlier revision. I'm fine with how it is for now, but in the future we may want to add a max-in-flight or something. I guess at the end of the day slow consumers are hard to avoid, and this is why we've been thinking of standardizing on Prometheus. :sweat_smile:

Reply:
@vito We have not had any issues with this, but I understand your concerns. As a final test I would suggest that once we have merged this and built a release candidate, that we will spin up a concourse cluster similar to our production one and perform some load tests to give us a measure of confidence with all the 5.5 code changes.